### DISCUSSION TITLE
Testing a Database not working at all

### DISCUSSION THREAD

**Discussion by pausanchezv** [fastapi/fastapi#8928] [Questions] [answered]
Hi,

What is claimed in this page of your documentation is not true: https://fastapi.tiangolo.com/advanced/testing-database/
In no way is this tutorial accurate. It just does not work, no matter what I do, it won't work at all.

After following this tutorial, my pytests pass only once. Why? because they still use the actual postgres database, and, in no way does the database get overridden by the testing one. This is really annoying, and even more after not having had an answer or solution to this problem when I opened a ticket for this before. 

Please, fix the explanations for this:
https://fastapi.tiangolo.com/advanced/testing-database/

Thanks.

**Comment by Kludex** [comment]
> What is claimed in this page of your documentation is not true: https://fastapi.tiangolo.com/advanced/testing-database/
In no way is this tutorial accurate. It just does not work, no matter what I do, it won't work at all.

There is no absolute truth in this world. :raised_hands: 

I was able to run that example just fine:
```bash
git clone https://github.com/tiangolo/fastapi
pytest docs_src/sql_databases/sql_app/tests
```

You can read more about pytest [here](https://docs.pytest.org/en/latest/how-to/usage.html). 

What is your other issue? Do you mind linking it here?

**Reply by Kludex** [reply]
> There is no absolute truth in this world. 🙌

3 years later... This is fun. hahahaha

**Comment by rotten** [comment]
I highly recommend this blog entry:  https://www.jeffastor.com/blog/testing-fastapi-endpoints-with-docker-and-pytest   it got us started on getting testing with fastapi working well.

**Comment by pausanchezv** [comment]
> > What is claimed in this page of your documentation is not true: https://fastapi.tiangolo.com/advanced/testing-database/
> > In no way is this tutorial accurate. It just does not work, no matter what I do, it won't work at all.
> 
> There is no absolute truth in this world. 🙌
> 
> I was able to run that example just fine:
> 
> ```shell
> git clone https://github.com/tiangolo/fastapi
> pytest docs_src/sql_databases/sql_app/tests
> ```
> 
> You can read more about pytest [here](https://docs.pytest.org/en/latest/how-to/usage.html).
> 
> What is your other issue? Do you mind linking it here?

There are no more issues, only this one. Will keep trying I guess.
Thanks for the rapid answer!

**Comment by pausanchezv** [comment]
> I highly recommend this blog entry: https://www.jeffastor.com/blog/testing-fastapi-endpoints-with-docker-and-pytest it got us started on getting testing with fastapi working well.

Thank you so much! This solution is really cool, although it doesn't use the native way of doing it (the one I'd like to use with dependency override).

My solution is close to yours, what I basically do is create and destroy a testing database myself, like this:
```
@pytest.fixture(scope="class", autouse=True)
def create_test_database():
    """
    Creates a clean testing database on every test class.
    Scope=class so that a database lasts an entire class and the objects created by the setup method will
    persist all the tests of that class. Then the tearDown method will drop the database and get ready for
    the next test.
    """
    if database_exists(app_config.database_url_testing):
        drop_database(app_config.database_url_testing)
    create_database(app_config.database_url_testing)

    # Auto-apply the database migrations
    Base.metadata.create_all(engine)

    # Run the tests
    yield

    # Doing away with the database when the tests have run
    drop_database(app_config.database_url_testing)
```

However, I really wanna use the fastapi way, but if it's impossible I'll survive with your approach and mine!
Thanks!

**Comment by pausanchezv** [comment]
> > > What is claimed in this page of your documentation is not true: https://fastapi.tiangolo.com/advanced/testing-database/
> > > In no way is this tutorial accurate. It just does not work, no matter what I do, it won't work at all.
> > 
> > 
> > There is no absolute truth in this world. 🙌
> > I was able to run that example just fine:
> > ```shell
> > git clone https://github.com/tiangolo/fastapi
> > pytest docs_src/sql_databases/sql_app/tests
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can read more about pytest [here](https://docs.pytest.org/en/latest/how-to/usage.html).
> > What is your other issue? Do you mind linking it here?
> 
> There are no more issues, only this one. Will keep trying I guess.
> Thanks for the rapid answer!

Just to clarify, I set a breakpoint inside of this function and it never gets there.

`def override_get_db():
    try:
        db = TestingSessionLocal()
        yield db
    finally:
        db.close()`

Which means that the following line does nothing
`app.dependency_overrides[get_db] = override_get_db`

I mean, I have other ways to use pytest creating and destroying a test database myself. However, since this is the native way to do it, I would like to take advantage of it. The problem is not the database creation since in this example it does get created. The problem is that fastapi still uses the production database.

**Comment by pausanchezv** [comment]
> > > > What is claimed in this page of your documentation is not true: https://fastapi.tiangolo.com/advanced/testing-database/
> > > > In no way is this tutorial accurate. It just does not work, no matter what I do, it won't work at all.
> > > 
> > > 
> > > There is no absolute truth in this world. 🙌
> > > I was able to run that example just fine:
> > > ```shell
> > > git clone https://github.com/tiangolo/fastapi
> > > pytest docs_src/sql_databases/sql_app/tests
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > You can read more about pytest [here](https://docs.pytest.org/en/latest/how-to/usage.html).
> > > What is your other issue? Do you mind linking it here?
> > 
> > 
> > There are no more issues, only this one. Will keep trying I guess.
> > Thanks for the rapid answer!
> 
> Just to clarify, I set a breakpoint inside of this function and it never gets there.
> 
> ```def override_get_db(): try: db = TestingSessionLocal() yield db finally: db.close()```
> 
> Which means that the following line does nothing
> `app.dependency_overrides[get_db] = override_get_db`
> 
> I mean, I have other ways to use pytest creating and destroying a test database myself. However, since this is the native way to do it, I would like to take advantage of it. The problem is not the database creation since in this example it does get created. The problem is that fastapi still uses the production database.

**Comment by pausanchezv** [comment]
> > > > > What is claimed in this page of your documentation is not true: https://fastapi.tiangolo.com/advanced/testing-database/
> > > > > In no way is this tutorial accurate. It just does not work, no matter what I do, it won't work at all.
> > > > 
> > > > 
> > > > There is no absolute truth in this world. 🙌
> > > > I was able to run that example just fine:
> > > > ```shell
> > > > git clone https://github.com/tiangolo/fastapi
> > > > pytest docs_src/sql_databases/sql_app/tests
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > You can read more about pytest [here](https://docs.pytest.org/en/latest/how-to/usage.html).
> > > > What is your other issue? Do you mind linking it here?
> > > 
> > > 
> > > There are no more issues, only this one. Will keep trying I guess.
> > > Thanks for the rapid answer!


Just to clarify, I set a breakpoint inside of this function and it never gets there.
```
def override_get_db():
    try:
        db = TestingSessionLocal()
        yield db
    finally:
        db.close()
```
Which means that the following line does nothing
`app.dependency_overrides[get_db] = override_get_db`
 I mean, I have other ways to use pytest creating and destroying a test database myself. However, since this is the native way to do it, I would like to take advantage of it. The problem is not the database creation since in this example it does get created. The problem is that fastapi still uses the production database.

**Comment by Acerinth** [comment, marked as answer]
Hey there 👋🏼 

Not sure if it helps, but... How I did it is just to define different settings in `conftest.py` for the database. E.g.
```python
SQLALCHEMY_DATABASE_URL = "sqlite:///./test.db"

class ForeignKeysListener(PoolListener):
    def connect(self, dbapi_con, con_record):
        db_cursor = dbapi_con.execute('pragma foreign_keys=ON')

engine = create_engine(
    SQLALCHEMY_DATABASE_URL, connect_args={"check_same_thread": False}, listeners=[ForeignKeysListener()]
)
TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

CoreBase.metadata.create_all(bind=engine)     # here the database gets created from ORM objects, but in the SQLite db

def override_get_db():
    try:
        db = TestingSessionLocal()
        yield db
    finally:
        db.close()

@pytest.fixture(scope="session")
def test_app():
    app.dependency_overrides[get_db] = override_get_db
    client = TestClient(app)
    yield client
```

P.S. In case you have submounts, you need to override this db in subapis separately as well. E.g.
```python
@pytest.fixture(scope="session")
def test_app():
    subapi.dependency_overrides[get_db] = override_get_db
    app.dependency_overrides[get_db] = override_get_db
    client = TestClient(app)
    yield client
```

Of course, this solution works if you use ORM - SQLAlchemy in my case - so you can just call `YourBase.metadata.create_all()`. If not... then I guess you need to implement table creation yourself. But you should still be able to at least override in which database tests happen anyways.
